Shifts : user type, skills, assignment draft

Theophile-Madet opened the pull request:
Hey,

This is a WIP of the shift feature. It has :

- A "my shifts" page that shows the user's shift type, skills, and assignments
- A rudimentary shift calendar. 

It's missing :
- translations
- tests

I thought I'd do a PR already to show that things are happening and to have feedback on that first draft so that I can apply it to the rest of the feature.

Theophile-Madet committed:
feat: added the shifts extension.

Theophile-Madet committed:
Added Qodana

Theophile-Madet committed:
Revert "Added Qodana"

This reverts commit d1eff25cf953d93e06c6da8876795202a2f7f1c8.

Theophile-Madet committed:
feat: shifts schema WIP

Theophile-Madet committed:
Added the option to specify the field type when creation O2M relations.
Prefixed all collections and fields names with the extension name.

Theophile-Madet committed:
Added example shift data and a link in the menu.

Theophile-Madet committed:
Added a simple shift overview page for a user.

Theophile-Madet committed:
Added assignments to shift examples.

Theophile-Madet committed:
Updated the shift schema to represent shift templates and occurences.

jofmi committed:
fix: run postinstall on shifts extension

Theophile-Madet committed:
WIP shifts

Theophile-Madet committed:
First draft of calculating shift occurences.

Theophile-Madet committed:
Draft for shift cards.

Theophile-Madet committed:
WIP Shift user type

Theophile-Madet committed:
WIP displaying assigments.

Theophile-Madet committed:
Improved assignment display.

Theophile-Madet committed:
WIP user <-> skills

Theophile-Madet committed:
Display skills on the shifts page.

Theophile-Madet committed:
Fixed shift calendar.

Theophile-Madet committed:
Typing improvements.

Theophile-Madet committed:
Added shift logs and score calculation.

Theophile-Madet committed:
Show last logs on the my shifts page.

Theophile-Madet committed:
WIP calendar view with FullCalendar.

Theophile-Madet committed:
Improved calendar display.

Theophile-Madet committed:
First draft for a shift detail page.

Theophile-Madet committed:
Moved the assignment list to a separate component.

jofmi commented on collectivo/extensions/shifts/LICENSE:
need to delete this, as the main license applies

jofmi commented on collectivo/extensions/shifts/composables/types.ts:
Why are these not in collectivo/extensions/shifts/index.d.ts?

jofmi commented on collectivo/extensions/shifts/pages/shifts/my_shifts.vue:
```diff
@@ -0,0 +1,114 @@
+const user = useCollectivoUser();
+await user.value.load();
+
+const assignments: CollectivoAssignment[] = await directus.request(
```
Can we have the database calls run in parallel? At the moment, if I read it right, they will run one after another before loading the component, right? Instead, we could remove the await keyword and add a v-if statement in the template to only load the component once all the relevant data is loaded.

Theophile-Madet replied:
Good point, updated 👍

jofmi commented on collectivo/extensions/shifts/components/AssignmentCard.vue:
```diff
@@ -0,0 +1,72 @@
+      <p>
+        Assigned to slot : {{ props.shiftAssignment.shifts_slot.shifts_name }}
+      </p>
+      <UAlert
```
How about instead of this alert, show the next occurance of a assignment in the card?

Theophile-Madet replied:
The date of the next occurrence is in the card's title actually. Maybe it would be clearer to have just the shift as title and the next occurrence separately inside the card?

Would you then remove the alert? I think it's still useful to make it clear when the next occurrence is not within the assignment.

jofmi replied:
Ah ok great! Let us leave the alert for now - as it is easier to remove it later than to add it back in :)

jofmi commented on collectivo/extensions/shifts/components/Calendar.vue:
I would move all Components in a folder /shifts - then Calendar will automatically be named ShiftsCalendar

jofmi commented on collectivo/extensions/shifts/components/Calendar.vue:
I cannot find where the onClick is defined, but it currently causes the whole app to reload. It might need to be changed to a NuxtLink. Or maybe it could also be nice to stay on the same page and show the shift details as a modal. What do you think?

Theophile-Madet replied:
Updated, I was giving a raw URL to FullCalendar, changed it to a custom click event so that I can use `navigateTo`

jofmi commented on collectivo/extensions/shifts/server/schemas/001_schema_shifts.ts:
```diff
@@ -0,0 +1,208 @@
+    field: "shifts_repeats_every",
+    type: "integer",
+    schema: { is_nullable: true },
+    meta: { note: "In days" },
```
i would add another field to make it an option whether to repeat each days/weeks/months - then the admin user could set it e.g. to "repeats every fourth week on monday", or every 1st monday of each month?

Theophile-Madet replied:
That would make more of what RRule offers available, that'd be nice, I'll update.

jofmi commented on collectivo/extensions/shifts/server/schemas/001_schema_shifts.ts:
```diff
@@ -0,0 +1,208 @@
+  },
+  {
+    collection: "shifts_shifts",
+    field: "shifts_from",
```
I think it could be more intuitive to have the shifts starting time as a separate field  and have this one only as a date (no time)

Theophile-Madet replied:
Makes sense the fields have a different meaning. I'll update.

jofmi commented on collectivo/extensions/shifts/server/schemas/001_schema_shifts.ts:
It is redundant from a data perspective, but for filtering tables it might be helpful to have a direct foreign key from slots & logs to the shift. What do you think?

Theophile-Madet replied:
According to https://docs.directus.io/app/data-model/relationships.html, we should be able to have the reverse link as a virtual field, so it would be convenient without being redundant in the DB. I'll have a look.

jofmi replied:
That sounds very interesting, also for other parts of collectivo! Let us see what you find :)

jofmi commented on collectivo/extensions/shifts/server/schemas/001_schema_shifts.ts:
```diff
@@ -0,0 +1,208 @@
+  },
+];
+
+schema.createO2MRelation("shifts_slots", "shifts_shifts", "shifts_shift");
```
Here we need to add a field on the shifts_shift side, so that you can edit the slots when you edit a shift.

jofmi commented on collectivo/extensions/shifts/pages/shifts/shift_calendar.vue:
What does the color of the dot next to the shift mean?

jofmi replied:
We'll need a bunch of filters and legends for shift type, shift active, week (A,B,C,D), etc.

Theophile-Madet replied:
> What does the color of the dot next to the shift mean?

It's from FullCalendar, currently it doesn't have a meaning.
We'll probably need a custom event template to show how full each occurrence is.

jofmi commented on collectivo/extensions/shifts/pages/shifts/my_shifts.vue:
Action flows to start/choose and end assignments is needed

jofmi reviewed (commented):
I think it looks really good already - the basic structure of what we need is already there and exactly as envisioned!

Theophile-Madet committed:
Removed the license file.

Theophile-Madet committed:
Avoid "await" calls in setup scripts, instead let the calls load in the background.

Theophile-Madet committed:
Moved all components to a shifts/ subfolder.

Theophile-Madet committed:
In the shift calendar, use navigateTo instead of a raw URL in order to avoid reloading the nuxt app when clicking a shift.

Theophile-Madet committed:
Renamed types from Collectivo* to Shifts* and added them to CollectivoSchema.

Theophile-Madet committed:
Cleaned up shift examples.

Theophile-Madet committed:
WIP show required skills on shift page.

Theophile-Madet committed:
WIP show required skills on shift page.

Theophile-Madet committed:
Show slot names and required skills on shift page.

Theophile-Madet committed:
Added assignee name to slots in shifts detail page.

Theophile-Madet committed:
Fixed missing icons after rebase from main repo.

Theophile-Madet committed:
Fixed assignments on my_shifts page.

Theophile-Madet committed:
Show assignments on occurrence cards.

Theophile-Madet committed:
Improved logs display on my_shifts page.

Theophile-Madet committed:
Added status fields for shifts, slots and assignments. Updated frontend to take them into account.

Theophile-Madet committed:
WIP directus flow shift cycle start.

Theophile-Madet committed:
Added automatic removal of points when a shift cycle starts.

Theophile-Madet committed:
First draft for assignment page.

Theophile-Madet committed:
Added assignment editing.

Theophile-Madet committed:
Formatting fix.

Theophile-Madet committed:
Moved assignment edit to a modal.

Theophile-Madet committed:
Fixed typing errors on the shift page.

Theophile-Madet committed:
WIP assignment creation.

Theophile-Madet committed:
Improvements to assignment creation.

Theophile-Madet committed:
Added memberships as dependency (temporarilly, to be removed later)

Theophile-Madet committed:
Added system-uicons library for the shifts calendar.

Theophile-Madet committed:
Fixed assignments showing as current when the next occurrence is not within the assignment.

Theophile-Madet committed:
Show tooltip on slot card when it's not possible to assign yourself.

Theophile-Madet committed:
Update shift page after an assignment has been created.

Theophile-Madet committed:
WIP assignment edition improvement :
moved the form to a separate component used by both the edit and the create modal.
Prepared previewing the assignments for the coming occurrences.
Updated NuxtUI in order to use the size attribute of UDivider.

Theophile-Madet committed:
Updated AssignmentCreateModal.vue : check if at least one occurrence is included in the new assignment before allowing assignment creation.

Theophile-Madet committed:
Merged AssignmentEditModal.vue and AssignmentCreateModal.vue.

Theophile-Madet committed:
Split the shifts.ts into shifts and assignments.

Theophile-Madet committed:
WIP

Theophile-Madet committed:
Finished assignment creation & update fixes.

Theophile-Madet committed:
Added tests to assignments utils, moved assignments utils from composables/ to utils/

Theophile-Madet committed:
Renamed fromToOverlaps to doTimeIntervalsOverlap and added tests.
Small refactoring in getOccurrenceType

Theophile-Madet committed:
Improved display of relational fields in Directus.
